client/src/composables/useData.js

```diff
@@ -0,0 +1,110 @@
+import { ref, readonly, computed } from "vue";
+import { checkIfAnyValueOfObjectIsDefined } from "../helper/utils";
+
+const DEFAULT_LIMIT = 50;
+const LIMIT_OPTIONS = [10, 25, 50, 100];
+
+const instances = {};
+
+export default (apiEndpoint, apiExtension) => {
+  if (!apiEndpoint) throw "No endpoint defined for useData";
+
+  if (!instances[apiEndpoint])
+    instances[apiEndpoint] = createInstance(apiEndpoint, apiExtension);
+
+  return instances[apiEndpoint];
+};
+
+function createInstance(apiEndpoint, apiExtension) {
+  const data = ref([]);
+  const sortOptionsCache = ref(null);
+  const filterOptionsCache = ref(null);
+  const paramsCache = ref(null);
+  const limitCache = ref(DEFAULT_LIMIT);
+  const numberOfTotalEntries = ref(0);
+  const isLoading = ref(false);
+  const currentRange = ref({ start: 0, end: 0 });
+
+  // Getters
+  const filterActive = computed(() =>
+    checkIfAnyValueOfObjectIsDefined(filterOptionsCache.value)
+  );
+
+  // Mutations
+  const clearFilter = () => {
+    filterOptionsCache.value = null;
+    fetchData();
+  };
+
+  const sortDataBy = async (sortOptions) => {
+    sortOptionsCache.value = sortOptions;
+    fetchData();
+  };
+
+  const filterDataBy = (filterOptions) => {
+    //Check if any filter value was set
+    if (!Object.values(filterOptions).find((v) => !!v)) return;
+    filterOptionsCache.value = filterOptions;
+    fetchData();
+  };
+
+  // Actions
+  const fetchData = async ({ params, queries, limit, offset = 0 } = {}) => {
+    if (params) paramsCache.value = params;
+    if (queries) filterOptionsCache.value = queries;
+    if (offset < 0) offset = 0;
+    if (limit) limitCache.value = limit;
+    try {
```

Comment:
Es gibt zwar hier einen try/catch block, allerdings nur ein isLoading und keine errorMessage im return. Wer auch immer fetchData() aufruft hat keine Ahnung ob das erfolgreich war, oder nicht.

Außerdem wird fetchData() jetzt manchmal mit await benutzt und manchmal nicht. Was war die intention?

Ich habe in den anderen composables den try/catch weggelassen und beim verwenden vom fetchX() einen block drum gesetzt. Dann kann der caller selber entscheiden, wie er Fehler händelt.

Reply:
Habe alle await entfernt, bis auf in ResultView.
Da werden die Daten aus dem Request ja noch benötigt in einer BackReference in Remarks benötigt.